newrelic/hooks/mlmodel_sklearn.py

```diff
@@ -79,13 +90,130 @@ def _nr_wrapper_method(wrapped, instance, args, kwargs):
             return PredictReturnTypeProxy(return_val, model_name=_class, training_step=training_step)
         return return_val
 
     wrap_function_wrapper(module, "%s.%s" % (_class, method), _nr_wrapper_method)
 
 
+def _calc_prediction_feature_stats(prediction_input, _class, feature_column_names):
+    import numpy as np
+
+    # Drop any feature columns that are not numeric since we can't compute stats
+    # on non-numeric columns.
+    x = np.array(prediction_input)
+    isnumeric_features = np.apply_along_axis(isnumeric, 0, x)
+    numeric_features = x[isnumeric_features]
+
+    # Drop any feature column names that are not numeric since we can't compute stats
+    # on non-numeric columns.
+    feature_column_names = feature_column_names[isnumeric_features[0]]
+
+    # Only compute stats for features if we have any feature columns left after dropping
+    # non-numeric columns.
+    num_cols = len(feature_column_names)
+    if num_cols > 0:
+        # Boolean selection of numpy array values reshapes the array to a single
+        # dimension so we have to reshape it back into a 2D array.
+        features = np.reshape(numeric_features, (len(numeric_features) // num_cols, num_cols))
+        features = features.astype(dtype=np.float64)
+
+        _record_stats(features, feature_column_names, _class, "Feature")
+
+
+def _record_stats(data, column_names, _class, column_type):
+    import numpy as np
+
+    mean = np.mean(data, axis=0)
+    percentile25 = np.percentile(data, q=0.25, axis=0)
+    percentile50 = np.percentile(data, q=0.50, axis=0)
+    percentile75 = np.percentile(data, q=0.75, axis=0)
+    standard_deviation = np.std(data, axis=0)
+    _min = np.min(data, axis=0)
+    _max = np.max(data, axis=0)
+    _count = data.shape[0]
+
+    transaction = current_transaction()
+
+    # Currently record_metric only supports a subset of these stats so we have
+    # to upload them one at a time instead of as a dictionary of stats per
+    # feature column.
+    for index, col_name in enumerate(column_names):
+        metric_name = "MLModel/Sklearn/Named/%s/Predict/%s/%s" % (_class, column_type, col_name)
+        transaction.record_custom_metrics(
+            [
+                ("%s/%s" % (metric_name, "Mean"), float(mean[index])),
+                ("%s/%s" % (metric_name, "Percentile25"), float(percentile25[index])),
+                ("%s/%s" % (metric_name, "Percentile50"), float(percentile50[index])),
+                ("%s/%s" % (metric_name, "Percentile75"), float(percentile75[index])),
+                ("%s/%s" % (metric_name, "StandardDeviation"), float(standard_deviation[index])),
+                ("%s/%s" % (metric_name, "Min"), float(_min[index])),
+                ("%s/%s" % (metric_name, "Max"), float(_max[index])),
+                ("%s/%s" % (metric_name, "Count"), _count),
+            ]
+        )
+    return (mean, percentile25, percentile50, percentile75, standard_deviation, _min, _max, _count)
```

Comment:
Do we need to return all our stats values here? I see two calls to this function but they don't appear to be assigning the return value to a variable that is used later.

Reply:
Ooh no we don't! Good catch!